### DISCUSSION TITLE
#172 のエラー調査

### DISCUSSION THREAD

**Discussion by kagomen** [lgtm-factory/lgtm-factory#180] [Q&A] [answered]
#172 で発生しているエラーを調査します

**Comment by kagomen** [comment]
コンソール見ると「POST https://update-top-page-166.lgtm-factory.pages.dev/ 405 (Method Not Allowed)」
ってなってる
まず、私ってPOSTしてたの・・・？となっている

**Reply by kagomen** [reply]
> Behind the scenes, actions use the POST method, and only this HTTP method can invoke them.

裏では、アクションはPOSTメソッドを使い、このHTTPメソッドだけがアクションを呼び出すことができる。

https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations#behavior

へー

**Comment by kagomen** [comment]
> 将来的に、APIを実装する際は、エッジランタイムを使用する必要があります。
（Cloudflareでは、現在はNext.jsの「Edge」ランタイムしか使用できません。）
4.すべてのサーバーレンダリングルートがエッジランタイムを使用していることを確認してください
[参考](https://developers.cloudflare.com/pages/framework-guides/nextjs/ssr/get-started/#4-ensure-all-server-rendered-routes-use-the-edge-runtime)
このため、サーバーサイドでの処理を実装する際には、そのファイルにて、
以下のような設定をする必要があることに注意してください！
export const runtime = "edge";

これかな？

**Reply by kagomen** [reply]
付け足したらエラーになった

> APIを実装する際は

と書いてあるし、route.tsxで必要なだけかも

**Reply by kagomen** [reply]
https://zenn.dev/microcms/articles/1b4331eca6e512#edge%E3%83%A9%E3%83%B3%E3%82%BF%E3%82%A4%E3%83%A0%E3%81%AE%E8%A8%AD%E5%AE%9A

> Cloudflare PagesではビルドはNode.jsが利用されますが、ランタイムははEdge Runtimeとなるので、利用するpage.tsxやroute.tsそれぞれにruntime設定を追記しておきます。

**利用するpage.tsx**にも？

**Reply by kagomen** [reply]
page.tsxに`export const runtime = "edge";`つけたら無限のエラーが出た

GET / 500 in 3908ms
 GET / 500 in 119ms
 GET /_next/static/webpack/82373d31059cf731.webpack.hot-update.json 500 in 66ms
 ⚠ Fast Refresh had to perform a full reload due to a runtime error.

**Reply by kagomen** [reply]
rootはSSGだから無理ですよってことなのかな
なんか「ランタイム」がいつ発動してるのか、いまいちわかってない

```
▲  Route (app)                              Size     First Load JS
▲  ┌ ○ /                                    13.6 kB         113 kB
▲  ├ ○ /_not-found                          141 B          87.2 kB
▲  ├ ƒ /api/v1/lgtm-images                  0 B                0 B
▲  ├ ○ /apple-icon.png                      0 B                0 B
▲  ├ ○ /icon.svg                            0 B                0 B
▲  └ ○ /manifest.json                       0 B                0 B
▲  + First Load JS shared by all            87.1 kB
▲  ├ chunks/23-9570260eafd46082.js        31.5 kB
▲  ├ chunks/fd9d1056-c4050f5b61ad0a3b.js  53.6 kB
▲  └ other shared chunks (total)          1.92 kB
▲  ○  (Static)   prerendered as static content
▲  ƒ  (Dynamic)  server-rendered on demand
```

**Reply by kagomen** [reply]
actions/getDesignInfo.tsに`export const runtime = "edge";`付けると、以下のエラーが出る

```
 ⨯ ./src/actions/getDesignInfo.ts
Error: 
  × Only async functions are allowed to be exported in a "use server" file.
   ╭─[/Users/fuka/Desktop/dev/lgtm-factory/src/actions/getDesignInfo.ts:1:1]
 1 │ "use server";
 2 │ 
 3 │ export const runtime = "edge";
   · ──────────────────────────────
 4 │ 
 5 │ import { ERRORS } from "@/app/api/v1/lgtm-images/_constants";
 6 │ import { GetLgtmData } from "@/types/lgtm-data";
   ╰────
```

**Reply by kagomen** [reply]
https://ja.next-community-docs.dev/docs/app-router/api-reference/edge

**Reply by kagomen** [reply]
`use server`と`export ~~ edge`が一緒に使えないのは、こういう理由っぽい

---

このエラーは、`"use server"`ディレクティブが含まれているファイル内で、非同期関数ではないもの（例えば`export const runtime = "edge";`）をエクスポートしようとしたために発生しています。

**原因**：
Next.jsのServer Actionsは、`"use server"`ディレクティブを使うことでサーバーサイドで実行されることを明示しますが、その場合、そのファイル内では**非同期関数**のみがエクスポート可能です。`export const runtime = "edge";`のように非同期ではない定数をエクスポートしようとすると、このエラーが発生します。

**解決方法**：
`runtime`を指定したい場合、そのファイルを`server`ファイルとして扱わず、別の場所でエクスポートするか、もしくは`getDesignInfo.ts`ファイルをエッジで実行する必要がない場合は、`runtime`のエクスポートを削除することができます。

もし、どうしても`runtime`を指定したい場合、以下のように分けることができます：

```typescript
// getDesignInfo.ts (Server Actions用)
"use server";
import { ERRORS } from "@/app/api/v1/lgtm-images/_constants";
import { GetLgtmData } from "@/types/lgtm-data";

// ここにServer Actionsの非同期関数を記述

// runtime.ts
export const runtime = "edge";
```

これで、`runtime`は`getDesignInfo.ts`内に存在しないため、エラーを回避できます。

**Comment by kagomen** [comment]
以下のページの内容をAIに食わせてみた

https://developers.cloudflare.com/pages/framework-guides/nextjs/ssr/supported-features/#rendering-and-data-fetching

---

はい、この情報は今回のエラーと関係がある可能性が高いです。Cloudflare Pages上でNext.jsを使用する際の制限事項について説明しています。主なポイントは以下の通りです：

1. Incremental Static Regeneration (ISR)の制限:
   Cloudflareは ISR をフルサポートしていません。静的フォールバックファイルは生成されますが、動的な再生成は行われません。

2. 動的ルートの制限:
   `@cloudflare/next-on-pages` は標準的な静的生成ルートをサポートしていますが、Node.jsベースの動的なオンデマンドハンドリングはサポートしていません。

3. `generateStaticParams` と `getStaticPaths` の制限:
   これらの関数の動作が制限される可能性があります。

4. キャッシュと再検証:
   `next/cache` とデータの再検証はサポートされていますが、Cloudflare特有の設定が必要な場合があります。

これらの制限を考慮すると、あなたのアプリケーションで以下のような問題が発生している可能性があります：

1. サーバーサイドの関数 (`getDesignInfo`) が期待通りに動作していない。
2. 動的なデータフェッチングが制限されている。
3. サーバーサイドのロジックがクライアントサイドで実行されようとしている。

解決策として、以下のアプローチを検討してください：

1. 静的生成の活用:
   可能な限り、ビルド時にデータを静的に生成し、JSONファイルとして保存する方法を検討してください。

2. クライアントサイドのデータフェッチング:
   サーバーサイドの関数に依存するのではなく、クライアントサイドで直接APIを呼び出す方法に変更してください。

3. Cloudflare Workersの利用:
   動的なデータ処理が必要な場合、Cloudflare Workersを使用してAPIを実装することを検討してください。

4. キャッシュ戦略の見直し:
   Cloudflareのキャッシュ機能を活用し、パフォーマンスを最適化することを検討してください。

5. SSRの使用:
   動的なコンテンツが必要な場合は、ISRの代わりにServer-Side Rendering (SSR) を使用することを検討してください。

これらの変更を適用することで、Cloudflare Pages上でのNext.jsアプリケーションの動作を改善できる可能性があります。また、Cloudflareの公式ドキュメントを参照して、特定の機能の制限や代替手段について詳細を確認することをお勧めします。

**Reply by kagomen** [reply]
> 'use server' は、クライアントサイドのコードから呼び出せる、サーバサイドの関数をマークします。

https://ja.react.dev/reference/rsc/use-server

**Reply by kagomen** [reply]
この機能を、cloudflareがサポートしているのか、というのがとても気になる

**Reply by kagomen** [reply]
エラーの原因わからなかったらAPIつくれば、まあ動きはするはず

**Reply by kagomen** [reply]
なにもわからないけど、

- cloudflareのデータフェッチングの欄にこんな感じのことが書いてあること
- next.jsのサーバーアクションの記事が、データフェッチングの記事に内包されていること

から、なんか関係がありそうだなという感じがする

**Reply by kagomen** [reply]
## Next.jsのサーバーアクションとデータフェッチングの関係

Next.jsのサーバーアクションは、データフェッチングと密接な関係があります。従来のデータフェッチングの方法と比較し、サーバーアクションがどのようにデータフェッチに影響を与えるのか、具体的な例を交えて解説します。

### 従来のデータフェッチングとサーバーアクションの違い

従来のNext.jsにおけるデータフェッチングは、主に以下の方法で行われてきました。

* **getStaticProps:** ビルド時にデータを取得し、静的なHTMLを生成する。
* **getServerSideProps:** リクエストごとにデータを取得し、動的なHTMLを生成する。
* **useSWR:** クライアントサイドでデータを取得し、キャッシュ機能も利用する。

これらの方法では、データの取得ロジックはコンポーネント内に記述されることが多く、コンポーネントの責務が肥大化しやすいという課題がありました。

**サーバーアクション**は、この課題を解決するために導入された機能です。サーバーアクションは、以下の特徴を持ちます。

* **コンポーネントの外にデータフェッチのロジックを分離できる:** フォーム送信やボタンクリックなどのイベントに紐づけて、サーバーサイドでデータの処理を行うことができます。
* **サーバーサイドレンダリング(SSR)と親和性が高い:** サーバーアクションはSSRと組み合わせて利用することで、SEO対策や初回表示速度の向上に貢献します。
* **フォーム処理に最適:** フォーム送信時のデータバリデーションやデータベースへの保存などを、サーバーサイドで安全に行うことができます。

### サーバーアクションとデータフェッチングの連携

サーバーアクションは、データフェッチングと以下のように連携します。

1. **クライアントサイドでのイベントトリガー:** フォーム送信やボタンクリックなどのイベントが発生します。
2. **サーバーサイドへのリクエスト:** イベントの情報がサーバーに送信されます。
3. **サーバーアクションの実行:** サーバーアクションが実行され、必要なデータが取得されます。
4. **クライアントサイドへのレスポンス:** 取得したデータや処理結果がクライアントに返されます。
5. **クライアントサイドでのUI更新:** 返されたデータに基づいて、UIが更新されます。

### サーバーアクションのメリット

* **コンポーネントの責務の明確化:** データフェッチのロジックをコンポーネントから分離することで、コンポーネントの可読性が向上します。
* **セキュリティの向上:** サーバーサイドでデータの処理を行うため、クライアントサイドでのJavaScriptの改ざんによるセキュリティリスクを軽減できます。
* **ユーザーエクスペリエンスの向上:** サーバーサイドレンダリングと組み合わせることで、初回表示速度が向上し、ユーザーエクスペリエンスが改善します。
* **フォーム処理の簡素化:** フォーム送信時のデータバリデーションやデータベースへの保存などを、サーバーアクションで一括して処理できます。

### まとめ

Next.jsのサーバーアクションは、データフェッチングの効率化とセキュリティの向上に大きく貢献する機能です。特に、フォーム処理やSSRとの連携においてその真価を発揮します。サーバーアクションを効果的に活用することで、より高品質なNext.jsアプリケーションを開発することができます。

### 具体的なユースケース

* **フォーム送信時のデータバリデーション:** サーバーサイドで厳密なデータチェックを行い、不正なデータの入力防止に役立ちます。
* **データベースへのデータ保存:** フォーム送信時に取得したデータをデータベースに安全に保存できます。
* **動的なページの生成:** リクエストごとに異なるデータを取得し、動的なページを生成できます。
* **認証処理:** ユーザーの認証情報をサーバーサイドで検証し、安全な認証処理を実現できます。

**Comment by kagomen** [comment]
https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations

> [Server Actions](https://react.dev/reference/rsc/server-actions) are asynchronous functions that are executed on the server. They can be called in Server and Client Components to handle form submissions and data mutations in Next.js applications.

サーバーアクションは、サーバー上で実行される非同期関数です。 Next.jsアプリケーションのフォーム送信やデータ変異を処理するために、サーバーコンポーネントやクライアントコンポーネントで呼び出すことができます。

**Reply by kagomen** [reply]
> Usually, the Next.js TypeScript plugin would flag updateItemAction in client-component.tsx since it is a function which generally can't be serialized across client-server boundaries. However, props named action or ending with Action are assumed to receive Server Actions. This is only a heuristic since the TypeScript plugin doesn't actually know if it receives a Server Action or an ordinary function. Runtime type-checking will still ensure you don't accidentally pass a function to a Client Component.

通常、Next.jsのTypeScriptプラグインは、client-component.tsxのupdateItemActionにフラグを立てます。 しかし、actionという名前のpropやActionで終わるpropは、Server Actionsを受け取るものと仮定されます。 TypeScriptプラグインは実際にServer Actionを受け取るのか、それとも普通の関数を受け取るのかを知らないため、これはヒューリスティックなものに過ぎない。 ランタイムの型チェックによって、誤ってClient Componentに関数を渡してしまうことがないようにすることができます。

**Reply by kagomen** [reply]
> Server Actions inherit the [runtime](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes) from the page or layout they are used on.

サーバー・アクションは、使用されるページやレイアウトのランタイムを継承します。

**Reply by kagomen** [reply]
https://github.com/lgtm-factory/lgtm-factory/discussions/159#discussioncomment-10500171

> pagesファイルでも、export const runtime = "edge";の記述をしたら、nodeモジュールは動かないですよ！
そうでなければ、静的ページとしてビルド時に生成されます！
ビルドは、node.jsなので使える感じですね👍

**Reply by kagomen** [reply]
Next.jsのサーバーアクションは、使用されているページやレイアウトがどのランタイム（Node.jsランタイムまたはEdgeランタイム）でレンダリングされているかに応じて、同じランタイムを継承します。

具体的には、次のような意味があります：

- **Node.jsランタイムでレンダリングされるページやレイアウト**でサーバーアクションを使用すると、そのサーバーアクションもNode.jsランタイム上で実行されます。Node.jsランタイムは、従来のサーバーサイドのNode.js環境を想定しており、豊富なライブラリやパッケージを使用できます。

- **Edgeランタイムでレンダリングされるページやレイアウト**でサーバーアクションを使用すると、そのサーバーアクションはEdgeランタイム上で実行されます。Edgeランタイムは、グローバルに分散されたサーバーレス環境を提供し、より高速なレスポンスを目指したものです。

これにより、サーバーアクションをどちらのランタイムでも使うことができ、必要に応じてパフォーマンスやデプロイメントの柔軟性を活用できます。

**Comment by kagomen** [comment]
## サーバーアクションの設定項目

https://nextjs.org/docs/app/api-reference/next-config-js/serverActions

> [allowedOrigins](https://nextjs.org/docs/app/api-reference/next-config-js/serverActions#allowedorigins)
> サーバーアクションを呼び出すことができる、特別に安全なオリジンドメインのリストです。 Next.jsはサーバーアクションリクエストのオリジンとホストドメインを比較し、それらが一致することを確認してCSRF攻撃を防ぎます。 指定されていない場合は、同じオリジンのみが許可されます。

**指定されていない場合は、同じオリジンのみが許可されます。**　とあるので、関係なさそう

あと他の項目は、全く関係なさそう

**Comment by kazzyfrog** [comment]
サーバーアクションではなく、データ取得用のAPIを生やしたほうが早いかもですね！

**Reply by kagomen** [reply]
@kazzyfrog 
ですよねー……
悔しいです！！！！

**Reply by kazzyfrog** [reply]
まあでも、毎回サーバーで処理するより、APIの方が、パフォーマンス的にも向上しそうですよ！

**Reply by kagomen** [reply]
そうなんですか……？👀
とりあえず、他に解決方法が見つからないので、APIルートで実装し直します🐈‍⬛

**Reply by kazzyfrog** [reply]
不思議な感じですが、Next.jsは、APIも静的に生成できるんです！

なので、たとえば、`api/v1/[theme]/route.tsx`があったとして、
ファイル内で、[theme]の動的な値に応じて、ダイナミックインポートし、デザインデータ・画像を返すAPIを定義した場合

それらは、いつアクセスされても、毎回返す内容が変わらないので、ビルド時に静的生成することができます👍

言い換えれば、
現在の画像生成APIが、アクセスごとに、テーマやクエリパラメータによって、動的な値で毎回サーバーで処理しているのに対し、
テーマも動的な値も固定した場合の別に立てたAPIのレスポンスは、ビルド時に生成すれば、毎回サーバーでの処理を行う必要がない感じです！

**Reply by kagomen** [reply]
@kazzyfrog 

なるほどーー！！

今回はcloudflareにデプロイしてるから、サーバーで処理するものには`export ~~ edge`を記述する必要があるけど、もし仮にvercelにデプロイするなら、完全に自動で区別してくれる感じなんでしょうか？？👀

**Reply by kazzyfrog** [reply]
それは確か、vercelでも、明示的に書いた方が良かった気がします！

**Reply by kagomen** [reply]
なるほど……🤨

**Comment by kagomen** [comment]
サーバーアクションよりも`route.tsx`でAPIを定義する方がパフォーマンスに優れる理由について詳しく説明します。

## **概要**

- **サーバーアクション**: React Server Components（RSC）の一部として導入された機能で、フォームやインタラクションから直接サーバーサイドの処理を呼び出すことができます。開発者体験を向上させ、コードの簡潔さを提供します。
- **`route.tsx`でのAPI定義**: 従来のRESTfulなAPIエンドポイントを定義する方法で、クライアントから明示的に`fetch`リクエストを送信し、サーバーサイドで処理を行います。

それぞれのアプローチには利点と欠点があり、特定のシナリオでは`route.tsx`でAPIを定義する方がパフォーマンス面で有利になる場合があります。

---

## **サーバーアクションとは**

### **特徴**

- **シンプルなAPI呼び出し**: サーバーアクションは、コンポーネント内で直接サーバーサイドの関数を呼び出すことができ、フォームの`onSubmit`ハンドラとして使用できます。
- **自動的なシリアライズ/デシリアライズ**: データの送受信において、Next.jsがデータのシリアライズとデシリアライズを自動的に処理します。
- **型安全性**: TypeScriptと組み合わせることで、型安全なAPI呼び出しが可能です。
- **開発者体験の向上**: コードがシンプルになり、フロントエンドとバックエンドの境界が曖昧になることで、開発効率が向上します。

### **例**

```jsx
// app/actions.js
"use server";

export async function createUser(data) {
  // サーバーサイドでユーザーを作成する処理
  // データベース操作など
}
```

```jsx
// app/page.jsx
import { createUser } from './actions';

export default function Page() {
  const handleSubmit = async (formData) => {
    await createUser(formData);
  };

  return (
    <form action={handleSubmit}>
      <input name="name" />
      <button type="submit">Create User</button>
    </form>
  );
}
```

---

## **`route.tsx`でのAPI定義とは**

### **特徴**

- **明示的なエンドポイント**: RESTfulなエンドポイントを定義し、クライアントから`fetch`リクエストを送信してデータを取得または送信します。
- **柔軟なレスポンス制御**: レスポンスのステータスコードやヘッダー、キャッシュ戦略などを細かく制御できます。
- **キャッシュの最適化**: CDNやブラウザキャッシュを効果的に活用し、レスポンスを高速化できます。
- **デバッグと監視の容易さ**: ネットワークリクエストが明示的に行われるため、デバッグやパフォーマンス監視が容易です。

### **例**

```tsx
// app/api/user/route.ts
import { NextRequest, NextResponse } from 'next/server';

export async function POST(request: NextRequest) {
  const data = await request.json();
  // ユーザーを作成する処理
  return NextResponse.json({ success: true }, { status: 201 });
}
```

```jsx
// app/page.jsx
export default function Page() {
  const handleSubmit = async (event) => {
    event.preventDefault();
    const res = await fetch('/api/user', {
      method: 'POST',
      body: JSON.stringify({ name: event.target.name.value }),
      headers: { 'Content-Type': 'application/json' },
    });
    const result = await res.json();
    // 結果の処理
  };

  return (
    <form onSubmit={handleSubmit}>
      <input name="name" />
      <button type="submit">Create User</button>
    </form>
  );
}
```

---

## **パフォーマンスの違いとその理由**

### **1. **ネットワークリクエストのオーバーヘッド**

- **サーバーアクション**:
  - サーバーアクションを使用する際、フォームの送信は特殊な形式で処理され、Next.jsが内部的にシリアライズとデシリアライズを行います。
  - このプロセスには追加のオーバーヘッドがあり、特に大きなデータや複雑なデータ構造を扱う場合にパフォーマンスに影響を与える可能性があります。
  - また、サーバーアクションはReactの内部プロトコルを使用しており、標準的なHTTPプロトコルと比べて最適化が不足している場合があります。

- **`route.tsx`でのAPI定義**:
  - 標準的なHTTPリクエストを使用するため、ネットワークスタックやブラウザ、CDNなどで最適化されています。
  - HTTP/2やHTTP/3の恩恵を受けることで、リクエストの多重化やヘッダーの圧縮などが行われ、パフォーマンスが向上します。
  - また、必要に応じてストリーミングや部分的なデータ取得などの高度な最適化も可能です。

### **2. キャッシュ戦略と再利用性**

- **サーバーアクション**:
  - 各リクエストごとに新しいレスポンスが生成され、キャッシュの制御が難しい場合があります。
  - 特に、同じデータを複数のユーザーが要求する場合、毎回サーバーサイドで処理が行われるため、サーバーの負荷が増加します。

- **`route.tsx`でのAPI定義**:
  - レスポンスヘッダーを詳細に設定することで、ブラウザやCDNでのキャッシュを効果的に利用できます。
  - 例えば、`Cache-Control`ヘッダーを設定して、静的なデータをクライアントサイドでキャッシュし、不要なサーバーリクエストを削減できます。
  - また、データの再利用性が高まり、ユーザー体験が向上します。

### **3. エラーハンドリングと再試行**

- **サーバーアクション**:
  - エラーハンドリングがブラックボックス化される傾向があり、詳細なエラー情報の取得や再試行ロジックの実装が難しい場合があります。
  - ネットワーク障害やサーバーエラーが発生した場合の対処が限定的になる可能性があります。

- **`route.tsx`でのAPI定義**:
  - ステータスコードやエラーメッセージを明示的に設定でき、クライアントサイドでのエラーハンドリングが容易です。
  - 再試行ロジックやバックオフ戦略を実装することで、信頼性の高い通信を実現できます。

### **4. セキュリティとアクセス制御**

- **サーバーアクション**:
  - サーバーアクションは、特定のコンポーネントやページに密接に結びついており、アクセス制御の実装が複雑になる場合があります。
  - 認証や認可のロジックを統一的に管理するのが難しいことがあります。

- **`route.tsx`でのAPI定義**:
  - APIごとにミドルウェアを設定し、認証や認可のロジックを

**Comment by kagomen** [comment]
実装はAPIでやるとして、

## 自分の理解をまとめてみる 

server actionsの挙動

- 開発環境(vercel)では動いたが、本番環境(cloudflare)では動かなかった
- server actions を cloudflare で使用する際は`export const runtime = 'edge`が必要？（ここ理解が怪しい）
- しかし page.tsx では fs を使用している
- fs を使用できているのは 、SSR がビルドによって SG として生成されているから（？）
- vercel だと node が動くけど、edge では node が動かない=fs モジュールなどは使えない
- これと衝突し、エラーが発生する

- MainSectionでfsモジュールを使ってる=SG
- MainSectionの子のImageInfoModalでserver actionsを使いたい、という状況
- ただし、ImageInfoModalはクライアントコンポーネントなので、`export const runtime = 'edge`が使えない
- じゃあ、ImageInfoModalを、クライアントコンポーネントとサーバーコンポーネントに分けてみる

**Reply by kagomen** [reply]
うーん　無理かも

**Comment by kagomen** [comment]
## APIに処理を変更したらビルドエラーが発生したので、原因を探る

build時のエラー内容
```
Generating static pages (7/8)  [=== ]SyntaxError: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
    at JSON.parse (<anonymous>)
    at parseJSONFromBytes (node:internal/deps/undici/undici:4306:19)
    at successSteps (node:internal/deps/undici/undici:4288:27)
    at fullyReadBody (node:internal/deps/undici/undici:2724:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async consumeBody (node:internal/deps/undici/undici:4297:7)
    at async X (/Users/fuka/Desktop/dev/lgtm-factory/.next/server/app/page.js:35:4319) {
  digest: '3513599283'
}
```

開発環境ではちゃんと動く

**Reply by kagomen** [reply]
- 環境変数
- レスポンスの返し方
- fetchの使い方

をもう一度確認する

**Comment by kagomen** [comment]
## 環境変数

全部読む
https://nextjs.org/docs/app/building-your-application/configuring/environment-variables

**Reply by kagomen** [reply]
> next/envで環境変数をロードする ORMやテストランナーのルート設定ファイルなど、Next.jsランタイムの外部で環境変数をロードする必要がある場合、@next/envパッケージを使用できます。 このパッケージは、Next.jsが内部的に.env*ファイルから環境変数をロードするために使用します。 使用するには、パッケージをインストールし、loadEnvConfig関数を使用して環境変数をロードします：

うーん、next buildでエラー出てるんだから関係なさそう

というか、next devで通ってnext buildでエラー出てるの謎だー

**Reply by kagomen** [reply]
> 知っておくとよいこと Next.jsアプリケーションをVercelにデプロイするとき、.env*ファイル内の環境変数は、NEXT_PUBLIC_というプレフィックスが付かない限り、Edge Runtimeでは利用できません。 環境変数はプロジェクト設定で管理することを強くお勧めします。

**Reply by kagomen** [reply]
読んだけど問題なさそうだった……

**Comment by kagomen** [comment]
## レスポンスの返し方

https://nextjs.org/docs/app/api-reference/functions/next-response

> NextResponse は Web レスポンス API に便利なメソッドを追加して拡張しています。

**Reply by kagomen** [reply]
> 指定されたJSONボディを持つレスポンスを生成する。

```
import { NextResponse } from 'next/server'
 
export async function GET(request: Request) {
  return NextResponse.json({ error: 'Internal Server Error' }, { status: 500 })
}
```

**Comment by kagomen** [comment]
エラー文に書いてあった
よむ

Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error

**Reply by kagomen** [reply]
> このエラーが発生した理由 ページのプリレンダリング中にエラーが発生しました。 このエラーは、ページフォルダにページ以外のコンポーネントなどを追加したり、propsに入力されるはずのものが入力されなかったりするなど、さまざまな理由で発生します。

**Reply by kagomen** [reply]
> 修正可能な方法
Next.js 13（またはそれ以降）とApp Routerを使用する。
すべての動的ページのpropsにデフォルト値を設定する（undefinedを避け、代わりにnullを使用してシリアライズできるようにする） 
依存している可能性のある古いモジュールがないか確認する getStaticPathsで有効になっている場合、コンポーネントがフォールバックを処理することを確認する。
 Fallback docs サーバーサイドレンダリングが有効になっているページをエクスポート (出力: 'export' または next export) しようとしていないことを確認してください 
getServerSideProps 必要であれば、window のようなブラウザ API を使おうとしている特定のコンポーネントに対して、サーバーサイドレンダリングを無効にすることができます。

**Reply by kagomen** [reply]
AIに聞いてみた
とりあえず言われた通りやってみる

---

このエラーは、サーバーサイドレンダリング（SSR）またはビルド時の静的生成（SSG）中に発生しているようです。主な問題は、コンポーネント内で非同期の`fetch`リクエストを行っていることです。

Next.jsのApp Routerでは、デフォルトでサーバーコンポーネントを使用しますが、このコンポーネントはクライアントサイドの機能（例：`Sheet`コンポーネント）も使用しているため、混乱が生じています。

解決策として、以下のアプローチを試してみてください：

1. このコンポーネントをクライアントコンポーネントとして明示的に宣言する：
   ファイルの先頭に `'use client';` を追加します。

2. データフェッチングロジックを別のサーバーコンポーネントに移動し、このコンポーネントにpropsとして渡す：

```jsx
// ImageInfoModalWrapper.tsx (Server Component)
import ImageInfoModal from './ImageInfoModal';

async function ImageInfoModalWrapper({ theme, children }) {
  let info;
  try {
    const response = await fetch(
      `${siteMetadata.SITE_URL}/api/v1/design-info?theme=${theme}`,
    );
    const { designInfo } = await response.json();
    info = designInfo;
  } catch (error: unknown) {
    if (error instanceof Error) {
      console.error(`fetch error: ${error.message}`);
      info = null;
    }
  }

  return <ImageInfoModal theme={theme} info={info}>{children}</ImageInfoModal>;
}

export default ImageInfoModalWrapper;

// ImageInfoModal.tsx (Client Component)
'use client';

// ... (他のインポート)

function ImageInfoModal({
  children,
  theme,
  info,
}: {
  children: React.ReactNode;
  theme: string;
  info: any;
}) {
  const url = `${siteMetadata.SITE_URL}/api/v1/lgtm-images?theme=${theme}`;

  return (
    <Sheet>
      {/* ... (残りのコンポーネントコード) */}
    </Sheet>
  );
}

export default ImageInfoModal;
```

3. `fetch`リクエストをクライアントサイドで行うように変更する：
   `useEffect`フックを使用してコンポーネントがマウントされた後にデータを取得します。

これらの変更により、ビルド時のエラーを回避し、コンポーネントが正しく動作するようになるはずです。

**Reply by kagomen** [reply]
エラーは取れたけど、やっぱりdesignInfoが表示されない
けど、fetchの部分が原因だということがわかったので、fetchの使い方を調査する

**Comment by kagomen** [comment]
https://ja.next-community-docs.dev/docs/app-router/building-your-application/routing/route-handlers#%E3%82%AD%E3%83%A3%E3%83%83%E3%82%B7%E3%83%A5%E3%81%AE%E3%82%AA%E3%83%97%E3%83%88%E3%82%A2%E3%82%A6%E3%83%88

> キャッシュのオプトアウト
キャッシュ無効にするには以下の方法があります：
GET メソッドで Request オブジェクトを使用する。
その他の HTTP メソッドを使用する。
cookies や headers などの [Dynamic Functions](https://ja.next-community-docs.dev/docs/app-router/building-your-application/routing/route-handlers#dynamic-functions) を使用する。
[セグメント設定オプション](https://ja.next-community-docs.dev/docs/app-router/building-your-application/routing/route-handlers#segment-config-options) で手動で動的モードを指定する。

え、GETメソッドでrequestオブジェクト使ってる
そもそも現在の実装方法って意図してないものになってるのかな（静的にAPIを生成できてない？）

**Reply by kagomen** [reply]
> 同様に、POSTメソッドは Route Handler を動的に評価します。

**Reply by kagomen** [reply]
> Edge and Node.js Runtimes
Route ハンドラは、ストリーミングのサポートを含め、Edge と Node.js の両方のランタイムをシームレスにサポートする同型の Web API を持っています。Route ハンドラは Pages や Layouts と同じ[route segment configuration](https://ja.next-community-docs.dev/docs/app-router/api-reference/file-conventions/route-segment-config)を使うので、汎用の[静的に再生成される](https://ja.next-community-docs.dev/docs/app-router/building-your-application/data-fetching/fetching-caching-and-revalidating#%E3%83%87%E3%83%BC%E3%82%BF%E3%81%AE%E5%86%8D%E6%A4%9C%E8%A8%BC)Route ハンドラのような待望の機能をサポートします。

何を言ってるのか全くわからない

**Comment by kagomen** [comment]
大人しくuseEffect使えばバグ解決する気はする・・・でも、実装方針としてそれで良いのかがわからない

**Reply by kagomen** [reply]
とりあえずやってみる！

**Reply by kagomen** [reply]
だめだ、同じエラーが発生した
エラー文がブラウザのコンソールで表示されるか、エディタのコンソールで表示されるのかの違いしかない

```
GET https://lgtm-factory.pages.dev/api/v1/design-info?theme=opengraph 404 (Not Found)
23-547baf2c39fe4cc5.js:1 fetch error: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
GET https://lgtm-factory.pages.dev/api/v1/design-info?theme=simple-emoji-browser 404 (Not Found)
23-547baf2c39fe4cc5.js:1 fetch error: Unexpected token '<', "<!DOCTYPE "... is not valid JSON
```

**Reply by kagomen** [reply]
でも、fetchがうまく行ってないことはわかる・・・
そして、開発環境の方はuseEffect使っても問題なかった

ん？
https://lgtm-factory.pages.dev/api/v1/design-info?theme=opengraph
に問い合わせて 404エラー出てるってことは、マージしたら↑のURLが存在することになるから、いけるってこと？

プレビュー環境と本番環境でURLが違ったのが原因だったってこと？

**Reply by kagomen** [reply]
だとしたら一つ前のコミットの、ImageInfoModalWrapper.tsxでfetchしてる実装もOKということになるかな

**Reply by kazzyfrog** [reply]
> だとしたら一つ前のコミットの、ImageInfoModalWrapper.tsxでfetchしてる実装もOKということになるかな

これは調査が必要ですが、多分いけるんじゃないかな？と思います！

単純に、ビルドでは、プロダクションのurlになるので、404が返っていただけ？かもです！

**Comment by kazzyfrog** [comment]
まず、`build`では、プロダクションのURLが適用されますよ！
ターミナルでも、「プロダクションビルド」的な表記が出るかと思うので！

なので、ローカルで`build`したら、
現状`main`からデプロイされている、プロダクションのAPIが叩かれるので、エラーになるかと！
（マージされていないので、404のHTMLが返されるので、[このエラー](https://github.com/lgtm-factory/lgtm-factory/discussions/180#discussioncomment-10528744)になる）

**Comment by kazzyfrog** [comment]
あと、`edge runtime`のAPIから、
`fetch()`で、ビルド時に静的にデータを取得することは、基本しないかなと思います！

（`edge runtime`のAPIは、ビルド時に叩けるのか？に関しては、不明なので、別で調査が必要ですが）

なので、静的に生成する実装をする場合、API自体を、`edge runtime`じゃなく、静的にする必要があるかもです！

**Reply by kagomen** [reply]
調べてみたら、Next.jsには`generateStaticParams`という関数を使って動的ルートのAPIを静的に生成する方法があるんですね！👀
これを使って書いてみます……！

**Reply by kazzyfrog** [reply, marked as answer]
おっしゃる通りです！
現状でも、問題なく動作すると思いますが、毎回サーバーにクエリが飛ぶ感じですね！

> `generateStaticParams`という関数を使って動的ルートのAPIを静的

上記と、下記の`export const dynamic = 'force-static'`を組み合わせることで、
静的なエンドポイントを、ビルド時に生成して、値をキャッシュできるかと思います👍
https://nextjs.org/docs/app/building-your-application/routing/route-handlers#caching

**Comment by kazzyfrog** [comment]
現状は、おそらく、`useEfeectで、毎回`edge runtime`のAPIを叩いて`fetch()`している実装ですが、
多分問題なくデータ取得できると思います！
